src/util/secrets/secrets.c

```diff
@@ -0,0 +1,1235 @@
+/*
+   SSSD
+
+   Local secrets database
+
+   Copyright (C) Red Hat 2018
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include "config.h"
+
+#include "util/util.h"
+#include "util/crypto/sss_crypto.h"
+#include "util/secrets/sec_pvt.h"
+#include "util/secrets/secrets.h"
+
+/* The KCM responder must "impersonate" the owner of the credentials.
+ * Only a trusted UID can do that -- root by default, but unit
+ * tests might choose otherwise */
+#ifndef KCM_PEER_UID
+#define KCM_PEER_UID            0
+#endif /* KCM_PEER_UID */
+
+#define MKEY_SIZE (256 / 8)
+
+#define SECRETS_BASEDN  "cn=secrets"
+#define KCM_BASEDN      "cn=kcm"
+
+#define LOCAL_SIMPLE_FILTER "(type=simple)"
+#define LOCAL_CONTAINER_FILTER "(type=container)"
+
+typedef int (*url_mapper_fn)(TALLOC_CTX *mem_ctx,
+                             const char *url,
+                             uid_t client,
+                             char **mapped_path);
+
+struct url_pfx_router {
+    const char *prefix;
+    url_mapper_fn mapper_fn;
+};
+
+static struct sss_sec_quota default_sec_quota = {
+    .max_secrets = DEFAULT_SEC_MAX_SECRETS,
+    .max_uid_secrets = DEFAULT_SEC_MAX_UID_SECRETS,
+    .max_payload_size = DEFAULT_SEC_MAX_PAYLOAD_SIZE,
+    .containers_nest_level = DEFAULT_SEC_CONTAINERS_NEST_LEVEL,
+};
+
+static struct sss_sec_quota default_kcm_quota = {
+    .max_secrets = DEFAULT_SEC_KCM_MAX_SECRETS,
+    .max_uid_secrets = DEFAULT_SEC_KCM_MAX_UID_SECRETS,
+    .max_payload_size = DEFAULT_SEC_KCM_MAX_PAYLOAD_SIZE,
+    .containers_nest_level = DEFAULT_SEC_CONTAINERS_NEST_LEVEL,
+};
+
+static int local_decrypt(struct sss_sec_ctx *sctx, TALLOC_CTX *mem_ctx,
+                         const char *secret, const char *enctype,
+                         char **plain_secret)
+{
+    char *output;
+
+    if (enctype && strcmp(enctype, "masterkey") == 0) {
+        DEBUG(SSSDBG_TRACE_INTERNAL, "Decrypting with masterkey\n");
+
+        struct sss_sec_data _secret;
+        size_t outlen;
+        int ret;
```

Comment:
Usually we're not declaring anything inside if-else blocks (although, personally I do believe it makes the code cleaner in general.
So, feel free to just ignore this one.

Reply:
I would prefer to not fix this, because that's how the code was written in the secrets responder, I just moved it.

Reply:
That's fine :-)